Add feed_success_score and other count tallies

benhalpern opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This is a follow up on other feed events work to close the loop on the summation step, and add this score as an option in feed experiments (though not included in any new live experiment).

While the task of tracking event details is a part of a broader future process of allowing us to better understand and tailer feed relevance, using this score summation to conduct a final ordering based on which posts are most likely to be well-liked (i.e. follow-through to a full reaction and/or comment) is a good step in this process.

## Related Tickets & Documents

- Related Issue #
- Closes https://github.com/forem/forem/issues/19926

## QA Instructions, Screenshots, Recordings

Tests should cover the general behavior, but gut check whether they are testing the right things.

## Added/updated tests?

- [x] Yes

benhalpern commented on app/javascript/articles/Feed.jsx:
```diff
@@ -377,4 +377,4 @@ Feed.propTypes = {
 };
 
-Feed.displayName = 'Feed';
+Feed.displayName = 'Feed';
```
```suggestion
Feed.displayName = 'Feed';

```

filleduchaos commented on app/models/feed_event.rb:
```diff
@@ -42,4 +47,30 @@ def self.record_journey_for(user, article:, category:)
+    reactions = article.feed_events.where(category: "reaction")
+    comments = article.feed_events.where(category: "comment")
+
+    distinct_impressions_score = impressions.distinct.count(:user_id)
+    distinct_clicks_count = clicks.distinct.count(:user_id)
+    reactions_score = reactions.distinct.count(:user_id) * REACTION_SCORE_MULTIPLIER
+    comments_score = comments.distinct.count(:user_id) * COMMENT_SCORE_MULTIPLIER
+
+    score = (distinct_clicks_count + reactions_score + comments_score) / distinct_impressions_score.to_f
```
So this is the real reason I mentioned rolling up events - less about saving raw storage space, and more that over time the _number_ of records has the potential to make these queries rather slow for what's supposed to be a fairly lightweight counter bump.

I think wrapping this in a `ThrottledCall` keyed per article ([as with billboard events](https://github.com/forem/forem/blob/c72c52dfdb3b2cea7ad69b11e904a8dbdef11879/app/controllers/billboard_events_controller.rb#L20-L32)) is fair for now, and we can revisit rolling up later

filleduchaos replied:
Oh whoops I forgot to mention - I'm a little iffy about the `distinct` clauses here, since they would drop a _lot_ of information. For example no matter how many times a user comes across an article in their feed, only one impression of theirs would ever count - although maybe that's the intention, but then that would make collecting data about article position and such (or even ever creating the extra impressions in the first place) rather unnecessary

benhalpern replied:
This is the intention strictly for this score calculation, which is sort of a fusion number meant to represent total outcomes. Definitely agree with you on the minute data, where we are not rolling up in this way. The _counts_, and individual data are full captures where we can retain all the data, or as far as we can do so efficiently.

filleduchaos reviewed (commented):
I thought through this a bit during the initial feed events implementation, so I do have some concerns

benhalpern commented:
@filleduchaos I adjusted the behavior to account for your concerns to the extent we can.

I added the throttle for a single article, and also added a bulk calculation which samples 5 events, in order to achieve a more reliable eventual consistency, while still limiting the total calls — both with the throttling and the sampling.

Let me know if this is a reasonable step towards the a good balance of consistency and efficiency.

filleduchaos commented on app/models/feed_event.rb:
```diff
@@ -42,4 +47,54 @@ def self.record_journey_for(user, article:, category:)
+  end
+
+  def self.update_single_article_counters(article_id)
+    ThrottledCall.perform("billboards_data_update-#{article_id}", throttle_for: 15.minutes) do
```
```suggestion
    ThrottledCall.perform(""article_feed_success_score_#{article_id}"", throttle_for: 15.minutes) do
```

maestromac replied:
@filleduchaos I think you accidentally included more double quotes here

filleduchaos replied:
😅 ah yes, thanks!

filleduchaos reviewed (approved):
🎉  I think we can revisit the specifics of the score calculation at a later time; it will likely take us a while and more data to settle on a decent algorithm anyway.

maestromac commented on app/services/feed_events/bulk_upsert.rb:
```diff
@@ -57,6 +57,7 @@ def call
       end
 
       FeedEvent.insert_all(records_to_insert) if records_to_insert.present?
+      FeedEvent.bulk_update_counters_by_article_id(records_to_insert.pluck(:article_id).sample(5))
```
I'm not sure if this matters, but the line before this accounts for when `if records_to_insert.present?` Should the same be done here?

benhalpern replied:
Good call.

maestromac commented on spec/models/feed_event_spec.rb:
```diff
@@ -63,4 +63,129 @@
+      # Reload the article to get the updated counters and scores
+      article.reload
+
+      expect(article.feed_success_score).to eq((1 + 5 + 10) / 2.0) # Calculated score
```
nit: May be worth to referring to the variable itself here, ie `FeedEvent::REACTION_SCORE_MULTIPLIER + FeedEvent::COMMENT_SCORE_MULTIPLIER +...` that way the test won't go stale easily.

benhalpern replied:
Good call

maestromac commented on spec/models/feed_event_spec.rb:
```diff
@@ -63,4 +63,129 @@
+    end
+
+    it "initializes counters and scores to zero when no events have occurred" do
+      # Trigger the after_save
+      create(:feed_event, category: "impression", article: article, user: user1)
+
+      article.reload
+
+      expect(article.feed_success_score).to eq(0.0)
+      expect(article.feed_impressions_count).to eq(1) # One impression
+      expect(article.feed_clicks_count).to eq(0) # Zero clicks
```
```suggestion
      expect do
        create(:feed_event, category: "impression", article: article, user: user1)
      end.to not_change { article.reload.feed_success_score }
        .and not_change { article.reload.feed_clicks_count }
        .and change { article.reload.feed_impressions_count }.from(0).to(1)

```
To test callbacks effectively, you have to verify the value before and after, else it may get tampered by other test. This is the most succinct way. It has the benefit of self-explanation of what should be changing and how much.

maestromac commented on app/models/feed_event.rb:
```diff
@@ -42,4 +47,54 @@ def self.record_journey_for(user, article:, category:)
+      score = (clicks_score + reactions_score + comments_score).to_f / distinct_impressions_users.size
+
+      # Update the article counters
+      Article.where(id: article_id).update_all(
+        feed_success_score: score,
+        feed_clicks_count: clicks.size,
+        feed_impressions_count: impressions.size,
+      )
```
question, are callbacks necessary here? Is it possible to `update_column` instead?

benhalpern replied:
`update_all` is one way here to not trigger callbacks, so ultimately a matter of taste in approach I think.

maestromac commented on app/models/feed_event.rb:
```diff
@@ -6,6 +6,8 @@ class FeedEvent < ApplicationRecord
   belongs_to :article, optional: true
   belongs_to :user, optional: true
 
+  after_save :update_article_counters_and_scores
```
I'm kind of out of the loop here. How often does this get saved? I see a throttle in place, but it's on a per-feed basis.

benhalpern replied:
It's only saved in some circumstances where a single record is created. In other scenarios it is bulk inserted, which won't trigger a full "save" as far as callbacks are concerned.

A single article is likely to get a lot of hits one way or another, so the throttling makes sense here I think.

benhalpern commented on app/models/articles/feeds.rb:
```diff
@@ -89,6 +89,13 @@ def self.lever_catalog
+                     order_by_fragment: "articles.feed_success_score DESC")
+
+      order_by_lever(:final_order_by_feed_success_score_and_primary_score,
+                     label: "Order by feed success score and score",
```
👍

benhalpern commented:
@maestromac thanks for all the feedback, made all requisite changes.

maestromac reviewed (approved):
LGTM!
